unity/Assets/Scripts/FifoServer.cs

```diff
@@ -0,0 +1,126 @@
+        private string serverPipePath;
+        private string clientPipePath;
+        private byte[] eomHeader;
+
+        
+        public static Dictionary<string, FieldType> FormMap = new Dictionary<string, FieldType>{
+            {"image", FieldType.RGBImage},
+            {"image_depth", FieldType.DepthImage},
+            {"image_normals", FieldType.NormalsImage},
+            {"image_classes", FieldType.ClassesImage},
+            {"image_flow", FieldType.FlowsImage},
+            {"image_ids", FieldType.IDsImage},
+            {"image-thirdParty-camera", FieldType.ThirdPartyCameraImage}
+
+        };
+
+        static public int UnpackNetworkBytes(byte[] data, int offset=0) {
+            int networkInt = System.BitConverter.ToInt32(data, offset);
+            return IPAddress.NetworkToHostOrder(networkInt);
+        }
+
+        static public byte[] PackNetworkBytes(int val) {
+            int networkInt = IPAddress.HostToNetworkOrder(val);
+            return System.BitConverter.GetBytes(networkInt);
+        }
+
+        public string ReceiveMessage() {
+            if (clientPipe == null) {
+                clientPipe = new NamedPipeClientStream(".", this.clientPipePath, PipeDirection.In);
+                clientPipe.Connect();
+            }
+            string action = null;
+            while (true) {
+                byte[] header = new byte[headerLength];
+                clientPipe.Read(header, 0, header.Length);
+                FieldType fieldType = (FieldType)header[0];
+                if (fieldType == FieldType.EndOfMessage) {
+                        //Console.WriteLine("Got eom");
+                    break;
+                }
+                int fieldLength = UnpackNetworkBytes(header, 1);
+                byte[] body = new byte[fieldLength];
+                clientPipe.Read(body, 0, body.Length);
+
+                switch (fieldType) {
+                    case FieldType.Action:
+                        //Console.WriteLine("Got action");
+                        action = Encoding.Default.GetString(body);
+                        break;
+                    default:
+                        throw new Exception("Invalid field type for Client: " + fieldType);
+                }
+            }
+            return action;
+        }
+
+        public void SendEOM() {
+            serverPipe.Write(this.eomHeader, 0, this.headerLength);
+            serverPipe.Flush();
+        }
+
+        public void SendMessage(FieldType t, byte[] body) {
+            if (this.serverPipe == null) {
+                this.serverPipe = new NamedPipeClientStream(".", this.serverPipePath, PipeDirection.Out);
+                this.serverPipe.Connect();
+            }
+            //Console.WriteLine("server pipe + connected " + this.serverPipe.IsConnected );
+            byte[] header = new byte[headerLength];
+            header[0] = (byte)t;
+            PackNetworkBytes(body.Length).CopyTo(header, 1);
```

Comment:
Why is the header copied to every location of the buffer starting at position `1` of the byte array?

Reply:
Each field is comprised of a HEADER + BODY.  The HEADER consists of a FIELD_TYPE + FIELD_LENGTH. The size of the HEADER is always 5 bytes (1 for the type and 4 (32 bits) for the length). So that is why we always have to read the header this way.

Reply:
Reading ReadMessage shouldn't this line be? 

`PackNetworkBytes(body.Length).CopyTo(body.Lenght, 1);`

or `IPAddress.HostToNetworkOrder(body.Lengh)` for correct endianness 

Like if I understand you are using the first byte to store the header `FieldType` and the rest of the bytes to store the integer representing the size of the body to be read.